browser-polyfill.js

```diff
@@ -0,0 +1,340 @@
+                target[prop] = value;
+              },
+            });
+
+            return value;
+          }
+
+          cache[prop] = value;
+          return value;
+        },
+
+        set(target, prop, value, receiver) {
+          if (prop in cache) {
+            cache[prop] = value;
+          } else {
+            target[prop] = value;
+          }
+          return true;
+        },
+
+        defineProperty(target, prop, desc) {
+          Object.defineProperty(cache, prop, desc);
+          return true;
+        },
+
+        deleteProperty(target, prop) {
+          delete cache[prop];
+          return true;
+        },
+      };
+
+      return new Proxy(target, handlers);
+    };
+
+    /**
+     * Creates a set of wrapper functions for an event object, which handles
+     * wrapping of listener functions that those messages are passed.
+     *
+     * A single wrapper is created for each listener function, and stored in a
+     * map. Subsequent calls to `addListener`, `hasListener`, or `removeListener`
+     * retrieve the original wrapper, so that  attempts to remove a
+     * previously-added listener work as expected.
+     *
+     * @param {DefaultWeakMap<function, function>} wrapperMap
+     *        A DefaultWeakMap object which will create the appropriate wrapper
+     *        for a given listener function when one does not exist, and retrieve
+     *        an existing one when it does.
+     *
+     * @returns {object}
+     */
+    const wrapEvent = wrapperMap => ({
+      addListener(target, listener, ...args) {
+        target.addListener(wrapperMap.get(listener), ...args);
+      },
+
+      hasListener(target, listener) {
+        target.hasListener(wrapperMap.get(listener));
+      },
+
+      removeListener(target, listener) {
+        target.removeListener(wrapperMap.get(listener));
+      },
+    });
+
+    const onMessageWrappers = new DefaultWeakMap(listener => {
+      if (typeof listener !== "function") {
+        return listener;
+      }
+
+      /**
+       * Wraps a message listener function so that it may send responses based on
+       * its return value, rather than by returning a sentinel value and calling a
+       * callback. If the listener function returns a Promise, the response is
+       * sent when the promise either resolves or rejects.
+       *
+       * @param {*} message
+       *        The message sent by the other end of the channel.
+       * @param {object} sender
+       *        Details about the sender of the message.
+       * @param {function(*)} sendResponse
+       *        A callback which, when called with an arbitrary argument, sends
+       *        that value as a response.
+       * @returns {boolean}
+       *        True if the wrapped listener returned a Promise, which will later
+       *        yield a response. False otherwise.
+       */
+      return function onMessage(message, sender, sendResponse) {
+        let result = listener(message, sender);
```

Comment:
Then it's dealt with the same way it would normally be dealt with if the listener throws. This version doesn't deal with propagating errors.